File: Python/S0005-longest-palindromic-substring.py
"""
Given a string s, return the longest palindromic substring in s.

Example 1:
Input: s = "babad"
Output: "bab"
Explanation: "aba" is also a valid answer.

Example 2:
Input: s = "cbbd"
Output: "bb"

Constraints:
1 <= s.length <= 1000
s consist of only digits and English letters.
"""


class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) == 0:
            return ""

        start = 0
        max_len = 1
        is_palindromic = []
        for i in range(len(s)):
            is_palindromic.append([False] * len(s))
        for i in range(len(s)):
            j = i
            while j >= 0:
                if (s[i] == s[j]) and ((i - j < 2) or (is_palindromic[j + 1][i - 1])):
                    is_palindromic[j][i] = True
                    if max_len < i - j + 1:
                        start = j
                        max_len = i - j + 1
                j -= 1
        return s[start:(start + max_len)]


if __name__ == '__main__':
    assert Solution().longestPalindrome("babad") == "bab"
    assert Solution().longestPalindrome("cbbd") == "bb"
    assert Solution().longestPalindrome("abccba") == "abccba"
    assert Solution().longestPalindrome("abcAcba") == "abcAcba"
    assert Solution().longestPalindrome("abcda") == "a"
    assert Solution().longestPalindrome("a") == "a"
    assert Solution().longestPalindrome("") == ""

File: Rust/S0005-longest-palindromic-substring.rs
// Given a string s, find the longest palindromic substring in s.
// You may assume that the maximum length of s is 1000.
//
// Example 1:
// Input: "babad"
// Output: "bab"
// Note: "aba" is also a valid answer.
//
// Example 2:
// Input: "cbbd"
// Output: "bb"
pub struct Solution {}

impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        let n = s.len();
        if n <= 1 {
            return s
        }
        let chars: Vec<char> = s.chars().collect();
        let (mut start, mut max_len) = (0, 1);
        let mut is_palindromic = vec![vec![false; n]; n];
        let mut j: usize = 0;
        for i in 0..n {
            j = i;
            loop {
                if chars[i] == chars[j] && (i-j<2 || is_palindromic[j+1][i-1]){
                    is_palindromic[j][i] = true;
                    if max_len < i-j+1 {
                        start = j;
                        max_len = i-j+1;
                    }
                }
                if j == 0 {
                    break;
                }
                j -= 1;
            }
        }
        return s[start..(start+max_len)].to_owned()
    }
}

fn main() {
    assert_eq!("bab".to_string(), Solution::longestPalindrome("babad".to_string()));
    assert_eq!("abccba".to_string(), Solution::longestPalindrome("cbbd".to_string()));
    assert_eq!("abccba".to_string(), Solution::longestPalindrome("abccba".to_string()));
    assert_eq!("abcAcba".to_string(), Solution::longestPalindrome("abcAcba".to_string()));
    assert_eq!("".to_string(), Solution::longestPalindrome("".to_string()));
    assert_eq!("a".to_string(), Solution::longestPalindrome("a".to_string()));
    println!("Pass test cases!");
}

